README.md
## jQuery PointerEvents v0.5.2 [![Build Status](https://secure.travis-ci.org/deepsweet/jquery-pointerevents.png)](https://travis-ci.org/deepsweet/jquery-pointerevents) [![Coverage Status](https://coveralls.io/repos/deepsweet/jquery-pointerevents/badge.png?branch=master)](https://coveralls.io/r/deepsweet/jquery-pointerevents?branch=master)

Basic polyfill for [Pointer Events W3C Specitication](https://dvcs.w3.org/hg/pointerevents/raw-file/tip/pointerEvents.html) using [jQuery special events API](http://www.benalman.com/news/2010/03/jquery-special-events/).

### what?

![PointerEvents](http://funkyimg.com/i/DWrJ.png)

Polyfill is trying to dynamically wrap all the possible `mouse*`, `touch*` and `MSPointer*` events to provide a single `PointerEvent` Interface. [Read more](https://dvcs.w3.org/hg/pointerevents/raw-file/tip/pointerEvents.html#intro).

### why?

Because there are too many different events types on different platforms with different input mechanisms. [Read more](http://docs.webplatform.org/wiki/concepts/Pointer_Events#Why_Pointer_Events).

### how?

```sh
$ bower install --save jquery-pointerevents
```

```html
<!-- load jQuery and plugin as browser globals or with RequireJS/AMD -->

<script>
    $('body').on({
        pointerdown: function(e) { console.log(e); },
        pointerup: function(e) { console.log(e); },
        // …
    });
</script>
```

### implemented

* `pointerenter`
* `pointerover`
* `pointerdown`
* `pointermove`
* `pointerup`
* `pointerout`
* `pointerleave`
* `pointercancel`

### not implemented

* `gotpointercapture`
* `lostpointercapture`
* `element.setPointerCapture()`
* `element.releasePointerCapture()`
* `onpointer*` HTML attributes
* `touch-action` CSS property
* `window.navigator.pointerEnabled`
* `window.maxTouchPoints`

### cross-browser compatibility

Depends on your jQuery version – [jquery.com/browser-support](http://jquery.com/browser-support/).

Also, [caniuse.com/#feat=pointer](http://caniuse.com/#feat=pointer):

```javascript
// nothing to do in IE11 for today
if(win.navigator.pointerEnabled) {
    return;
}
```

### test

* open `test/index.html` in your browser(s), do something and look at console output
* `npm test`

### alternatives

* [Polymer / PointerEvents](https://github.com/Polymer/PointerEvents)
* [HandJS](https://handjs.codeplex.com/)

### license

[MIT](https://github.com/deepsweet/jquery-pointerevents/blob/master/LICENSE)


CHANGELOG.md
### 0.5.2 / 22.01.2014
* fix simulated events relatedTarget
* fix events bubbling
* update `README` and `.gitignore`

### 0.5.1 / 29.11.2013
* fix preventing mouse ghosts
* fix simulated events real target
* minor codestyle fix

### 0.5.0 / 28.11.2013
* large code refactoring
* isolate all the events from pointermove bind
* fix trigger/bubbling order of events
* fix mouse events bubbling in some cases
* test: `-ms-touch-action: none` for IE10
* AMDify
* update `README`

### 0.4.0 / 26.11.2013
* fix all the known problems with events bubbling on `touch*`
* fix `msPointer*` events duplicates
* fix `normalizeTouchEvent` target
* large code refactoring
* move compiled files to `dist/`
* `CHANGELOG.md`

### 0.3.6 / 22.11.2013
* fix default `e.pointerType` regocnition

### 0.3.5 / 21.11.2013
* fix pointermove bubbling, again

### 0.3.4 / 14.11.2013
* fix pointermove bubbling
* fix multitouch detection
* tests with mocha

### 0.3.3 / 12.11.2013
* fix `e.pointerType` for `touch*`

### 0.3.2 / 12.11.2013
* export `$.PointerEvent` before IE11 return
* update `Makefile`

### 0.3.1 / 09.11.2013
* `Makefile`
* fix dist
* `header.js`
* fix `bower.json`

### 0.3.0 / 09.11.2013
* dist + min
* export `$.PointerEvent`
* update `README`

### 0.2.1 / 09.11.2013
* `pointerevents.jquery.json`
* remove unnecessary comment
* update JSDoc
* update `README`

### 0.2.0 / 08.11.2013
* init
* dynamically wrap all the possible events types

### 0.1.0
* [bem-core/…/jquery__event_type_pointers.js](https://github.com/bem/bem-core/blob/0e7714bdab3af1e5a2dae749fea91eb6148495ab/common.blocks/jquery/__event/_type/jquery__event_type_pointers.js)


test/test.js
describe('common:', function() {

    it('$.PointerEvent must exists in window', function() {
        window.$.PointerEvent.must.exist();
    });

    it('$.PointerEvent must exists in window even if windown.navigator.pointerEnabled', function() {
        if(!window.navigator.pointerEnabled) {
            window.navigator.pointerEnabled = 'test';
        }

        window.$.PointerEvent.must.exist();

        if(window.navigator.pointerEnabled === 'test') {
            delete window.navigator.pointerEnabled;
        }
    });

    it('must successfully unbind from all aliased events', function() {
        var spy = sinon.spy(),
            elem = $('<div/>');

        elem
            .on('pointerenter', spy).off('pointerenter').mouseenter()
            .on('pointerover', spy).off('pointerover').mouseover()
            .on('pointerdown', spy).off('pointerdown').mousedown()
            .on('pointermove', spy).off('pointermove').mousemove()
            .on('pointerup', spy).off('pointerup').mouseup()
            .on('pointerleave', spy).off('pointerleave').mouseleave()
            .on('pointerout', spy).off('pointerout').mouseout();

        spy.callCount.must.be.equal(0);
    });

});

describe('interface:', function() {

    it('must return valid PointerEvent', function() {
        var spy = sinon.spy(),
            elem = $('<div/>'),
            e = $.Event('mousedown', {
                which: 1
            }),
            args;

        elem.on('pointerdown', spy).trigger(e);

        spy.calledOnce.must.be.true();

        args = spy.args[0][0];
        args.isPrimary.must.be.true();
        args.pointerId.must.be.equal(1);
        args.pointerType.must.be.equal('mouse');
        args.pressure.must.be.equal(0.5);
        args.width.must.be.equal(0);
        args.height.must.be.equal(0);
        args.tiltX.must.be.equal(0);
        args.tiltY.must.be.equal(0);
    });

});

describe('mouse:', function() {

    it('must trigger "pointerenter" event on "mouseenter"', function() {
        var spy = sinon.spy(),
            elem = $('<div/>');

        elem.on('pointerenter', spy).mouseenter();

        spy.calledOnce.must.be.true();
    });

    it('must trigger "pointerover" event on "mouseover"', function() {
        var spy = sinon.spy(),
            elem = $('<div/>');

        elem.on('pointerover', spy).mouseover();

        spy.calledOnce.must.be.true();
    });

    it('must trigger "pointerdown" event on "mousedown"', function() {
        var spy = sinon.spy(),
            elem = $('<div/>');

        elem.on('pointerdown', spy).mousedown();

        spy.calledOnce.must.be.true();
    });

    it('must trigger "pointermove" event on "mousemove"', function() {
        var spy = sinon.spy(),
            elem = $('<div/>');

        elem.on('pointermove', spy).mousemove();

        spy.calledOnce.must.be.true();
    });

    it('must trigger "pointerup" event on "mouseup"', function() {
        var spy = sinon.spy(),
            elem = $('<div/>');

        elem.on('pointerup', spy).mouseup();

        spy.calledOnce.must.be.true();
    });

    it('must trigger "pointerleave" event on "mouseleave"', function() {
        var spy = sinon.spy(),
            elem = $('<div/>');

        elem.on('pointerleave', spy).mouseleave();

        spy.calledOnce.must.be.true();
    });

    it('must trigger "pointerout" event on "mouseout"', function() {
        var spy = sinon.spy(),
            elem = $('<div/>');

        elem.on('pointerout', spy).mouseout();

        spy.calledOnce.must.be.true();
    });

});

describe('touch:', function() {

    var fakeTouchObj = {
            clientX: 100,
            clientY: 200
        };

    it('must trigger "pointerenter" event on "touchstart"', function() {

        var spy = sinon.spy(),
            e = $.Event('touchstart', {
                touches: [],
                targetTouches: [fakeTouchObj],
                changedTouches: [fakeTouchObj]
            }),
            elem = $('<div/>');

        elem.on('pointerenter', spy).trigger(e);

        spy.calledOnce.must.be.true();

    });

    it('must trigger "pointerover" event on "touchstart"', function() {

        var spy = sinon.spy(),
            e = $.Event('touchstart', {
                touches: [],
                targetTouches: [fakeTouchObj],
                changedTouches: [fakeTouchObj]
            }),
            elem = $('<div/>');

        elem.on('pointerover', spy).trigger(e);

        spy.calledOnce.must.be.true();

    });

    it('must trigger "pointerdown" event on "touchstart"', function() {

        var spy = sinon.spy(),
            e = $.Event('touchstart', {
                touches: [],
                targetTouches: [fakeTouchObj],
                changedTouches: [fakeTouchObj]
            }),
            elem = $('<div/>');

        elem.on('pointerdown', spy).trigger(e);

        spy.calledOnce.must.be.true();

    });

});

describe('mixed sources:', function() {

    it('must trigger only one touch-based event in case of touch + mouse', function() {

        var spy = sinon.spy(),
            fakeTouchObj = {
                clientX: 100,
                clientY: 200
            },
            e = $.Event('touchstart', {
                touches: [],
                targetTouches: [fakeTouchObj],
                changedTouches: [fakeTouchObj]
            }),
            elem = $('<div/>'),
            args;

        elem.on('pointerdown', spy).trigger(e).mousedown();

        spy.calledOnce.must.be.true();

        args = spy.args[0][0];
        args.pointerType.must.be.equal('touch');
        args.clientX.must.be.equal(100);
        args.clientY.must.be.equal(200);

    });

    it('must trigger only one mouse-based event in case of touch + mouse + next-tick-mouse', function(done) {

        var spy = sinon.spy(),
            fakeTouchObj = {
                clientX: 100,
                clientY: 200
            },
            fakeMouseObj = {
                clientX: 300,
                clientY: 400
            },
            touchEvent = $.Event('touchstart', {
                touches: [],
                targetTouches: [fakeTouchObj],
                changedTouches: [fakeTouchObj]
            }),
            mouseEvent = $.Event('mousedown', fakeMouseObj),
            elem = $('<div/>'),
            args;

        elem.on('pointerdown', spy).trigger(touchEvent).trigger(mouseEvent);

        setTimeout(function() {
            elem.trigger(mouseEvent);

            spy.calledOnce.must.be.false();
            args = spy.args[1][0];
            args.pointerType.must.be.equal('mouse');
            args.clientX.must.be.equal(300);
            args.clientY.must.be.equal(400);

            done();
        }, 0);

    });

});


src/jquery-pointerevents.js
/**
 * Basic polyfill for Pointer Events W3C Specification.
 *
 * @author Kir Belevich <kir@soulshine.in>
 * @copyright Kir Belevich 2013
 * @license MIT
 * @version 0.5.1
 */
 (function(factory) {

    /*global define*/
    'use strict';

    // AMD
    if(typeof define === 'function' && define.amd) {
        define([ 'jquery' ], factory);
    // browser globals
    } else {
        factory(jQuery);
    }

 }(function($) {

    /*global jQuery*/
    'use strict';

    /*
       http://www.w3.org/TR/pointerevents/
       https://dvcs.w3.org/hg/pointerevents/raw-file/tip/pointerEvents.html
       https://dvcs.w3.org/hg/webevents/raw-file/default/touchevents.html
       http://msdn.microsoft.com/en-US/library/ie/hh673557.aspx
       http://www.benalman.com/news/2010/03/jquery-special-events/
       http://api.jquery.com/category/events/event-object/
    */

    var win = window,
        doc = win.document,
        binds = {
            mouse: {
                enter: 'mouseenter',
                over: 'mouseover',
                down: 'mousedown',
                move: 'mousemove',
                up: 'mouseup',
                out: 'mouseout',
                leave: 'mouseleave'
            },

            touch: {
                enter: 'touchstart',
                over: 'touchstart',
                down: 'touchstart',
                move: 'touchmove',
                up: 'touchend',
                out: 'touchend',
                leave: 'touchend',
                cancel: 'touchcancel'
            },

            mspointer: {
                over: 'MSPointerOver',
                down: 'MSPointerDown',
                move: 'MSPointerMove',
                up: 'MSPointerUp',
                out: 'MSPointerOut',
                cancel: 'MSPointerCancel'
            }
        };

    /**
     * Normalize touch-event by keeping all the
     * possible properties normalized by jQuery.
     *
     * @see http://api.jquery.com/category/events/event-object/
     *
     * @param {Object} e event
     */
    function normalizeTouchEvent(e) {

        if(e.pointerType === 'touch') {

            e.originalEvent = e.originalEvent || e;

            // multitouch
            if(e.originalEvent.touches.length > 1) {
                e.multitouch = true;
                return;
            }

            var touchPoint = e.originalEvent.changedTouches[0];

            // keep all the properties normalized by jQuery
            e.clientX = touchPoint.clientX;
            e.clientY = touchPoint.clientY;
            e.pageX = touchPoint.pageX;
            e.pageY = touchPoint.pageY;
            e.screenX = touchPoint.screenX;
            e.screenY = touchPoint.screenY;
            e.layerX = e.originalEvent.layerX;
            e.layerY = e.originalEvent.layerY;
            e.offsetX = e.layerX - e.target.offsetLeft;
            e.offsetY = e.layerY - e.target.offsetTop;
            e.identifier = touchPoint.identifier;
        }

    }

    /**
     * Extend event to match PointerEvent Interface.
     *
     * @see https://dvcs.w3.org/hg/pointerevents/raw-file/tip/pointerEvents.html#pointer-events-and-interfaces
     * @see https://dvcs.w3.org/hg/webevents/raw-file/default/touchevents.html
     *
     * @param {object} e event
     */
    function extendToPointerEvent(e) {

        /*eslint complexity:0*/
        e.width = e.width ||
                  e.webkitRadiusX ||
                  e.radiusX ||
                  0;

        e.height = e.width ||
                   e.webkitRadiusY ||
                   e.radiusY ||
                   0;

        // TODO: stupid Android somehow could send "force" > 1 ;(
        e.pressure = e.pressure ||
                     e.mozPressure ||
                     e.webkitForce ||
                     e.force ||
                     e.which && 0.5 ||
                     0;

        e.tiltX = e.tiltX || 0;
        e.tiltY = e.tiltY || 0;

        switch(e.pointerType) {
            case 2: e.pointerType = 'touch'; break;
            case 3: e.pointerType = 'pen'; break;
            case 4: e.pointerType = 'mouse'; break;
            default: e.pointerType = e.pointerType;
        }

        e.isPrimary = true;

        // "1" is always for mouse, so +2 because of touch can start from 0
        e.pointerId = e.identifier ? e.identifier + 2 : 1;

    }

    /**
     * Mutate an event to PointerEvent.
     *
     * @param {object} e current event object
     * @param {string} type future pointerevent type
     */
    function PointerEvent(e, type) {

        extendToPointerEvent(e);
        normalizeTouchEvent(e);
        e.type = type;

        $.extend(this, e);

    }

    // export PointerEvent class
    $.PointerEvent = PointerEvent;

    // nothing to do in IE11 for today
    if(win.navigator.pointerEnabled) {
        return;
    }

    /**
     * Simple nextTick polyfill.
     *
     * @see http://jsperf.com/settimeout-vs-nexttick-polyfill
     *
     * @return {Function}
     */
    function nextTick(callback) {

        var msgName = 'nextTick-polyfill',
            timeouts = [];

        if(win.nextTick) {
            return win.nextTick(callback);
        }

        if(!win.postMessage || win.ActiveXObject) {
            return setTimeout(callback, 0);
        }

        win.addEventListener('message', function(e){
            if(e.source === win && e.data === msgName) {
                if(e.stopPropagation) {
                    e.stopPropagation();
                }

                if(timeouts.length) {
                    timeouts.shift()();
                }
            }
        }, false);

        timeouts.push(callback);
        win.postMessage(msgName, '*');

    }

    /**
     * Create new $.event.special wrapper with some default behavior.
     *
     * @param {string} type event type
     * @param {object} toExtend object to extend default wrapper
     */
    function addPointerEvent(type, toExtend) {

        var eventName = 'pointer' + type,
            pointerevent,

            eventSpecial = $.event.special[eventName] = {
                // bind
                setup: function() {
                    $(this)
                        .on(binds.mouse[type], eventSpecial.mouseHandler)
                        .on(binds.touch[type], eventSpecial.touchHandler)
                        .on(binds.mspointer[type], eventSpecial.msHandler);
                },

                // unbind
                teardown: function() {
                    $(this)
                        .off(binds.mouse[type], eventSpecial.mouseHandler)
                        .off(binds.touch[type], eventSpecial.touchHandler)
                        .off(binds.mspointer[type], eventSpecial.msHandler);
                },

                // mouse
                mouseHandler: function(e) {
                    // do not duplicate PointerEvent if
                    // touch/mspointer is already processed
                    if(!eventSpecial._noMouse) {
                        e.pointerType = 4;
                        pointerevent = new PointerEvent(e, eventName);
                        $(e.currentTarget).triggerHandler(pointerevent);
                    }

                    // clear the "processed" key right after
                    // current event and all the bubblings
                    nextTick(function() {
                        eventSpecial._noMouse = false;
                    });
                },

                // touch
                touchHandler: function(e) {
                    // stop mouse events handling
                    eventSpecial._noMouse = true;

                    e.pointerType = 2;
                    pointerevent = new PointerEvent(e, eventName);

                    $(e.currentTarget).triggerHandler(pointerevent);
                },

                // mspointer
                msHandler: function(e) {
                    // stop mouse events handling
                    eventSpecial._noMouse = true;

                    pointerevent = new PointerEvent(e, eventName);
                    $(e.target).trigger(pointerevent);
                }
            };

        // extend this $.event.special wrapper
        if(toExtend) {
            $.extend(eventSpecial, toExtend({
                event: eventSpecial,
                name: eventName,
                type: type
            }));
        }

    }

    /**
     * Object to extend $.event.special to touchmove-based events.
     *
     * @param {object} params
     * @param {object} params.event event object
     * @param {string} params.name event name
     * @param {string} params.type event type
     * @return {object}
     */
    function touchmoveBased(params) {

        var event = params.event,
            type = params.type;

        return {
            // bind
            setup: function() {
                $(this)
                    .on(binds.mouse[type], event.mouseHandler)
                    .on(binds.touch[type], event.touchHandler)
                    .on(binds.touch.down, event.touchDownHandler)
                    .on(binds.mspointer[type], event.msHandler);

                if(type !== 'move') {
                    $(this).on(binds.touch.move, event.touchMoveHandler);
                }
            },

            // unbind
            teardown: function() {
                $(this)
                    .off(binds.mouse[type], event.mouseHandler)
                    .off(binds.touch[type], event.touchHandler)
                    .off(binds.touch.down, event.touchDownHandler)
                    .off(binds.mspointer[type], event.msHandler);

                if(type !== 'move') {
                    $(this).off(binds.touch.move, event.touchMoveHandler);
                }
            },

            touchDownHandler: function(e) {
                // stop mouse events handling
                event._noMouse = true;
                // save initial target
                event._target = e.target;
            }
        };

    }

    /**
     * Object to extend $.event.special to pointerenter.
     *
     * @param {object} params
     * @param {object} params.event event object
     * @param {string} params.name event name
     * @param {string} params.type event type
     * @return {object}
     */
    function extendToEnter(params) {

        return $.extend(touchmoveBased(params), {
            touchMoveHandler: function(e) {
                e.pointerType = 2;

                var pointerevent = new PointerEvent(e, params.name),
                    targetFromPoint = doc.elementFromPoint(
                        pointerevent.clientX,
                        pointerevent.clientY
                    ),
                    target = params.event._target;

                // new target
                if(target !== targetFromPoint) {
                    // fix simulated event targets
                    pointerevent.relatedTarget = pointerevent.target;
                    pointerevent.target = pointerevent.targetFromPoint;

                    // inner target
                    if(target.contains(targetFromPoint)) {
                        $(targetFromPoint).triggerHandler(pointerevent);
                    // truly new target
                    } else if(!targetFromPoint.contains(target)) {
                        $(targetFromPoint).trigger(pointerevent);
                    }

                    // targetFromPoint -> target
                    params.event._target = targetFromPoint;
                }
            }
        });

    }

    /**
     * Object to extend $.event.special to pointerover.
     *
     * @param {object} params
     * @param {object} params.event event object
     * @param {string} params.name event name
     * @param {string} params.type event type
     * @return {object}
     */
    function extendToOver(params) {

        return $.extend(touchmoveBased(params), {
            touchMoveHandler: function(e) {
                e.pointerType = 2;

                var pointerevent = new PointerEvent(e, params.name),
                    targetFromPoint = doc.elementFromPoint(
                        pointerevent.clientX,
                        pointerevent.clientY
                    ),
                    target = params.event._target;

                // new target
                if(target !== targetFromPoint) {
                    // fix simulated event targets
                    pointerevent.relatedTarget = pointerevent.target;
                    pointerevent.target = pointerevent.targetFromPoint;

                    $(targetFromPoint).trigger(pointerevent);

                    // targetFromPoint -> target
                    params.event._target = targetFromPoint;
                }
            }
        });

    }

    /**
     * Object to extend $.event.special touchHandler with "target from point".
     *
     * @param {object} params
     * @param {object} params.event event object
     * @param {string} params.name event name
     * @param {string} params.type event type
     * @return {object}
     */
    function extendWithTargetFromPoint(params) {

        return {
            touchHandler: function(e) {
                // stop mouse events handling
                params.event._noMouse = true;

                e.pointerType = 2;

                var pointerevent = new PointerEvent(e, params.name),
                    targetFromPoint = doc.elementFromPoint(
                        pointerevent.clientX,
                        pointerevent.clientY
                    );

                // fix simulated event targets
                pointerevent.relatedTarget = pointerevent.target;
                pointerevent.target = pointerevent.targetFromPoint;

                $(targetFromPoint).triggerHandler(pointerevent);
            }
        };

    }

    /**
     * Object to extend $.event.special to pointerout.
     *
     * @param {object} params
     * @param {object} params.event event object
     * @param {string} params.name event name
     * @param {string} params.type event type
     * @return {object}
     */
    function extendToOut(params) {

        return $.extend(
            touchmoveBased(params),
            extendWithTargetFromPoint(params),
            {
                touchMoveHandler: function(e) {
                    e.pointerType = 2;

                    var pointerevent = new PointerEvent(e, params.name),
                        targetFromPoint = doc.elementFromPoint(
                            pointerevent.clientX,
                            pointerevent.clientY
                        ),
                        target = params.event._target;

                    // new target
                    if(target !== targetFromPoint) {
                        $(target).trigger(pointerevent);

                        // targetFromPoint -> target
                        params.event._target = targetFromPoint;
                    }
                }
            }
        );

    }

    /**
     * Object to extend $.event.special to pointerleave.
     *
     * @param {object} params
     * @param {object} params.event event object
     * @param {string} params.name event name
     * @param {string} params.type event type
     * @return {object}
     */
    function extendToLeave(params) {

        return $.extend(
            touchmoveBased(params),
            extendWithTargetFromPoint(params),
            {
                touchMoveHandler: function(e) {
                    e.pointerType = 2;

                    var pointerevent = new PointerEvent(e, params.name),
                        targetFromPoint = doc.elementFromPoint(
                            pointerevent.clientX,
                            pointerevent.clientY
                        ),
                        target = params.event._target;

                    // new target
                    if(target !== targetFromPoint) {
                        if(targetFromPoint.contains(target)) {
                            $(target).triggerHandler(pointerevent);
                        } else {
                            $(e.currentTarget).triggerHandler(pointerevent);
                        }

                        // targetFromPoint -> target
                        params.event._target = targetFromPoint;
                    }
                }
            }
        );

    }

    /**
     * Object to extend $.event.special to pointermove.
     *
     * @param {object} params
     * @param {object} params.event event object
     * @param {string} params.name event name
     * @param {string} params.type event type
     * @return {object}
     */
    function extendToMove(params) {

        return $.extend(
            touchmoveBased(params),
            extendWithTargetFromPoint(params)
        );

    }

    // init pointer events
    addPointerEvent('enter', extendToEnter);
    addPointerEvent('over', extendToOver);
    addPointerEvent('down');
    addPointerEvent('move', extendToMove);
    addPointerEvent('up', extendWithTargetFromPoint);
    addPointerEvent('out', extendToOut);
    addPointerEvent('leave', extendToLeave);
    addPointerEvent('cancel');

 }));
